tools/mkldscript.c

```diff
@@ -1,298 +1,465 @@
-          "    .debug_frame    0 : { *(.debug_frame) }"                                     "\n"
-          "    .debug_str      0 : { *(.debug_str) }"                                       "\n"
-          "    .debug_loc      0 : { *(.debug_loc) }"                                       "\n"
-          "    .debug_macinfo  0 : { *(.debug_macinfo) }"                                   "\n"
-        // SGI/MIPS DWARF 2 extensions
-          "    .debug_weaknames 0 : { *(.debug_weaknames) }"                                "\n"
-          "    .debug_funcnames 0 : { *(.debug_funcnames) }"                                "\n"
-          "    .debug_typenames 0 : { *(.debug_typenames) }"                                "\n"
-          "    .debug_varnames  0 : { *(.debug_varnames) }"                                 "\n"
-        // DWARF 3
-          "    .debug_pubtypes 0 : { *(.debug_pubtypes) }"                                  "\n"
-          "    .debug_ranges   0 : { *(.debug_ranges) }"                                    "\n"
-        // DWARF Extension
-          "    .debug_macro    0 : { *(.debug_macro) }"                                     "\n"
-          "    .gnu.attributes 0 : { KEEP (*(.gnu.attributes)) }"                           "\n", fout);
-
-    // Discard all other sections not mentioned above
-    fputs("    /DISCARD/ :"                                                                 "\n"
-          "    {"                                                                           "\n"
-          "       *(*);"                                                                    "\n"
-          "    }"                                                                           "\n", fout);
-    fputs("}\n", fout);
+    fprintf(f, "\t_RomEnd = _RomSize;\n\n");
+
+    /* end */
+    write_sections_finale(f);
+    fprintf(f, "}\n");
+
+    fclose(f);
 }
 
 static void usage(const char *execname)
 {
-    fprintf(stderr, "Nintendo 64 linker script generation tool v0.03\n"
-                    "usage: %s SPEC_FILE LD_SCRIPT\n"
-                    "SPEC_FILE  file describing the organization of object files into segments\n"
-                    "LD_SCRIPT  filename of output linker script\n",
-                    execname);
+    fprintf(stderr,
+        "Ocarina of Time linker script generator\n"
+        "usage:\n"
+        "\t%s -s <spec-file> <segment-dir> <segment-name>\n"
+        "\t%s -r <spec-file> <segment-dir> <output-file>\n",
+        execname,
+        execname
+    );
 }
 
 int main(int argc, char **argv)
 {
-    FILE *ldout;
     void *spec;
     size_t size;
 
-    if (argc != 3)
+    if (argc == 5 && strcmp(argv[1], "-s") == 0)
     {
-        usage(argv[0]);
-        return 1;
-    }
+        struct Segment seg;
 
-    spec = util_read_whole_file(argv[1], &size);
-    parse_rom_spec(spec, &g_segments, &g_segmentsCount);
+        spec = util_read_whole_file(argv[2], &size);
```

Comment:
Purely for the tool's source readability, I think it would be good to assign `argv[i]` to local variables with proper names so it doesn't seem so random what gets passed to functions

For example `char* specFilePath = argv[2];`